ledger/store/store.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/hashicorp/go-multierror"
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	lru "github.com/optakt/golang-lru"
+)
+
+// NOTE: When loading a checkpoint, so many payloads need to be written at once that they, in most cases, get added to
+// the store faster than it can persist them on disk. When that is the case and that the LRU cache is full, each call to
+// the Save method becomes blocking until the previous entry has been evicted from the cache and persisted on disk,
+// which is effectively as slow as always writing on disk for each operation. Under normal operating conditions however,
+// this solution allows us to store payloads on disk with a negligible impact to insertion performance and a limited
+// impact to memory usage.
+
+// TODO: Implement an alternative version of the store that puts all payloads simply into a map[[32]byte]*Payload.
+
+// Store is a component that provides fast persistent storage by using an LRU cache from which evicted entries get
+// persisted on disk.
+type Store struct {
+	log zerolog.Logger
+
+	db   *badger.DB
+	sema *semaphore.Weighted
+	tx   *badger.Txn
+	txMu *sync.RWMutex   // guards the current transaction against concurrent access
+	wg   *sync.WaitGroup // keeps track of when the flush goroutine should exit
+	err  chan error
+
+	cache     *lru.Cache
+	cacheSize int
+	isCleanMu *sync.RWMutex     // guards the dirty tracking map against concurrent access
+	isClean   map[[32]byte]bool // keeps track of whether cached entries are on disk
+
+	done chan struct{}
+}
+
+// New creates a new store using a cache of the given size and storing payloads on disk at the given path.
+func New(log zerolog.Logger, opts ...Option) (*Store, error) {
+	logger := log.With().Str("component", "payload_store").Logger()
+
+	config := DefaultConfig
+	for _, opt := range opts {
+		opt(&config)
```

Comment:
Each option function sets fields in the `config`, yes. For example, if `DefaultConfig` has an attribute like `Address` set to `google.com` by default, and I pass `WithAddress("yahoo.com")` to this function, it will iterate on the given functions, find the returned function of `WithAddress` and call it on the config's pointer, which will then be modified so that `config.Address` is now `yahoo.com`. Does that make sense? 😄

Reply:
Yes.
What I meant was that the function name is not an action, but a noun

Reply:
ah, you mean `opt`/`option`? It's a common way to name functional options